### ISSUE TITLE
Add Single Body Device Model

### ISSUE THREAD

**Issue by H0R5E** [SNL-WaterPower/WecOptTool#123]
As discussed in our last call, it makes a lot of sense to add the ability to model a single body device (probably in heave). This will help us encapsulate code that is shared between all models, and which is shared between similar classes, such as bodies in heave. 

A proposed case study for this model was [WaveBot](https://mhkdr.openei.org/submissions/151), but others candidates may also be viable.

**Comment by ryancoe** [comment]
We were actually referring to the "[WaveBot](https://www.youtube.com/watch?v=XCB12Het4c4)"

**Comment by ryancoe** [comment]
@H0R5E - I'm looking now at how to do this. The math is quite easy (effectively a simplification from the RM3), but I need your guidance on the programming structure. Here's my understanding:

- `WecOptTool.RM3Study` gives the user interface methods
- `WecOptLib.models.DeviceModelTemplate` is some sort of standard interface
- `WecOptLib.models.RM3.*` contains data as well as functions (not methods) that calculate power, etc. as a file called `DeviceModel` that somehow points everything back to `WecOptLib.models.DeviceModelTemplate`

Seems like I will need to copy-paste and then edit the following:

1. `getPSCoefficients.m`
2. `RM3_BEM.mat`
3. `getHydrodynamics.m`
4. `getMotion.m`

Note that looking at `WecOptTool.RM3Study`, I don't think this would need to be edited, so perhaps we should change the name.

**Comment by H0R5E** [comment]
OK, yeah, so:

1. Copy the WecOptTool\toolbox\+WecOptLib\+models\+RM3 directory to +WaveBot
2. Modify as required...
3. To start change line 46 in WecOptTool\toolbox\+WecOptTool\run.m to `RM3Device = WecOptLib.models.WaveBot.DeviceModel();`

That will allow you to develop the model and we can think of a way to build in a switch and rename as appropriately (I will happily work on that bit, if you want). 

I suspect this is the only time we will define another device this way (its not really user friendly), but it will be useful to confirm whether the abstraction of the controllers proposed in #126 are genuine (i.e. they don't vary per device).

**Comment by H0R5E** [comment]
I think this really shows up the limitations of the current architecture, as the geometry options for RM3 make little sense for this device. This approach will be a sticking plaster to achieve our current goal, but very much not a recommended approach.

**Comment by H0R5E** [comment]
My current thoughts are that we abstract the Study class in WecOptTool so that we have something like this:

```
+WecOptTool.RM3.Study
               .geom.Parametric
                    .Scalar
                    .Existing
           .WaveBot.Study
                   .geom.Scalar
                        .Existing
+WecOptTool.control. ...
+WecOptTool.AbsStudy
...
```

Not nice, but fits with the 0.1 style structure before we can provide a more generic interface.

**Comment by H0R5E** [comment]
We should probably do this over two PRs. You set up the model and add a dirty config switch to run.m like `run(study, 'WaveBot') that will choose between the RM3 and WaveBot device models (in if statements). I will do another PR in parallel to set up the interface changes proposed above.

**Comment by ryancoe** [comment]
I was thinking about this more over the weekend. I think we may need to step back and reconsider the structure a bit more (or maybe you already have and I just need to catch up). Looking more at problem based optimization in MATLAB, I think we need to think about WecOptTool’s function as forming objective functions and problem structures. I don’t think this is a dramatic departure from the current structure. By relying as heavily as possible on the structure developed by MATHWORKS, we can spend more time on the WEC-specific work.

https://www.mathworks.com/help/optim/ug/problem-based-workflow.html

**Comment by H0R5E** [comment]
I think we need to have a phone call about this (with just me, yourself and maybe Sterling), so that I can understand more about how you want to apply this idea. I'm not sure I understand how this is much different from the solver based approach for our current black-box like structure, unless you want to expose the workings of the PS controller so that the user can reprogram the constraints. Is that the idea? If so, then yes, we really do need a big redesign because I was always working on the assumption that the controller implementation would be hidden (to the general user). Perhaps you could provide an example (even in pseudo-code) of how you see this being applied, to help me understand.

Whatever we decide, I don't think that the 0.2 target is where we should attempt this due to our time constraints and limited resources. Even if we were able to get it working, the documentation workload of exposing the implementation in WecOptLib for general use is very high and we will need to keep updating it as changes such as #126 take effect (if that is still applicable). If you don't want to make this example available to the public then we could ditch this release and just work towards the final structure (whatever that may be) and not worry about the documentation until we have something more stable.
